### PR TITLE
bug 1677457 - Implement the UUID type in the RLB

### PR DESCRIPTION
**Pull request by chutten**

### CODE REVIEW

**Comment by chutten** [comment]
May interfere with mdboom's #1328

File: glean-core/rlb/src/private/uuid.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    ///
+    /// * `error` - The type of error
+    /// * `ping_name` - represents the optional name of the ping to retrieve the
+    ///   metric for. Defaults to the first value in `send_in_pings`.
+    ///
+    /// # Returns
+    ///
+    /// The number of errors reported.
+    #[allow(dead_code)] // Remove after mozilla/glean#1328
+    fn test_get_num_recorded_errors<'a, S: Into<Option<&'a str>>>(
```
Comment (Dexterp37) [inline glean-core/rlb/src/private/uuid.rs:46]: Let's make this public :)

File: glean-core/src/traits/uuid.rs [hunk 2]
```diff
@@ -27,5 +27,5 @@ pub trait Uuid {
     ///
     /// * `ping_name` - represents the optional name of the ping to retrieve the
     ///   metric for. Defaults to the first value in `send_in_pings`.
-    fn test_get_value<'a, S: Into<Option<&'a str>>>(&self, ping_name: S) -> Option<String>;
+    fn test_get_value<'a, S: Into<Option<&'a str>>>(&self, ping_name: S) -> Option<uuid::Uuid>;
```
Comment (Dexterp37) [inline glean-core/src/traits/uuid.rs:30]: We should also add `test_get_num_recorded_errors` to the trait

File: glean-core/src/metrics/uuid.rs [hunk 3]
```diff
@@ -59,10 +59,9 @@ impl UuidMetric {
     /// # Arguments
     ///
     /// * `glean` - The Glean instance this metric belongs to.
-    pub fn generate_and_set(&self, storage: &Glean) -> Uuid {
+    pub fn generate_and_set(&self, storage: &Glean) {
```
Comment (Dexterp37) [inline glean-core/src/metrics/uuid.rs:62]: If we do this then this API suddenly becomes different with respect to the other bindings: we should either restore it here or change the other bindings, too (filing a bug should be ok if we don't have one already)

cc @badboy

**Review by Dexterp37 — CHANGES_REQUESTED** [review]

File: glean-core/src/metrics/uuid.rs (hunk 3 above)
Comment (chutten) [reply glean-core/src/metrics/uuid.rs:62]: Oops, right, I'd forgotten I'd only filed https://bugzilla.mozilla.org/show_bug.cgi?id=1673017 and it hadn't been processed/adopted. I definitely need to put this back for now.

I guess my Opinions were too strong when I wrote it. : )

**Comment by chutten** [comment]
RFAL

**Review by Dexterp37 — APPROVED** [review]
Looks good! Would you kindly add a changelog entry and appease the CI gods before merging?

File: CHANGELOG.md [hunk 4]
```diff
@@ -11,6 +11,7 @@
 * Rust
   * Introduce the Counter metric type in the RLB.
   * Introduce the String metric type in the RLB.
+  * Introduce the UUID metric type in the RLB.
```
Comment (mdboom) [inline CHANGELOG.md:14]: @chutten: Oops, I think this is in the wrong changelog section...

Comment (chutten) [reply CHANGELOG.md:14]: Oh bother.

Comment (chutten) [reply CHANGELOG.md:14]: https://github.com/mozilla/glean/pull/1333

**Pull request merged**